objax/module.py

```diff
@@ -81,26 +80,40 @@ def __getitem__(self, key: Union[int, slice]):
         return value
 
 
-class ModuleWrapper(Module):
-    """Module whose sole purpose is to store a collectable VarCollection. This class is exclusively
-    used internally by Objax, for example in Jit, Vectorize and Parallel."""
+class Function(Module):
+    """Turn a function into a Module by keeping the vars it uses."""
 
-    def __init__(self, vc: VarCollection):
-        super().__init__()
-        self.vc = VarCollection((f'({self.__class__.__name__}){k}', v) for k, v in vc.items())
-
-    def vars(self, scope: str = '') -> VarCollection:
-        """Collect all the variables (and their names) contained in the VarCollection.
+    def __init__(self, f: Callable, vc: VarCollection):
+        """Function constructor.
 
         Args:
-            scope: string to prefix to the variable names.
-        Returns:
-            A VarCollection of all the variables.
+            f: the function or the module to represent.
+            vc: the VarCollection of variables used by the function.
         """
-        return VarCollection((scope + k, v) for k, v in self.vc.items())
+        self.vc = VarCollection((f'{{{f.__name__}}}.{k}', v) for k, v in vc.items())
```

Comment:
FIY if `f` is declared as lambda function, then `f.__name__` will be just `<lambda>` which might be not very informative/useful to insert.
So if you have several functions declared as follows:
```
f1 = Function(lambda x, y: ((model(x) - y)**2).sum(), model_vars)
f2 = Function(lambda x, y: ..., model_vars)
```
then naming of the variables in var collection will be the same for both functions.

Another corner case when `f` is declared using partial: `f = functools.partial( .... )`, in such case `f.__name__` is undefined despite f being callable:
```
Python 3.8.5 (default, Jul 20 2020, 18:32:44) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import functools
>>> def f(a, b):
...   return a + b
... 
>>> f.__name__
'f'
>>> f2 = functools.partial(f, b=2)
>>> f2
functools.partial(<function f at 0x7f27596b9430>, b=2)
>>> f2.__name__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'functools.partial' object has no attribute '__name__'
>>> callable(f2)
True

```

Reply:
I understand the points you make but not their relevance. What do you suggest?

Reply:
either don't prepend anything to variable names (because it may not matter that much), or prepend something so any callable will work and ideally different callables could be better differentiated (not sure if it's possible though)

Reply:
1. Not prepending anything: it's less debuggable so there's no upside to it.
2. Prepend so that any callable would work: any callable does work regardless of prepending. It's unrelated.

Basically I prefer to stick to whatever Python does and that's why I designed it this way.

Reply:
As I mentioned above, `f2 = functools.partial(f1, arg1=val1, ...)` is a callable, but it will cause runtime error because there is not `f2.__name__`.

Basically following code will cause runtime error:

```python
def f1(arg1, arg2):
  # ...

f2 = functools.partial(f1, arg1=val1, ...)
f_module = Function(f2, vars)
```

Reply:
Thanks, I added code to handle this + test case.